### ISSUE TITLE
XML-RPC for web2 - propose API changes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10744]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#1132|
|Type|enhancement|
|Created|2005-08-06 00:14:45Z|

Attachments:

* [xmlrpc.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/081/081be7c370bf9e7b4c6e696276c1b2d57623b26b/7ab5f6045f4d7664349eaf74afb482a43414a7e8.py) (7393 bytes) - added by oubiwann on 2006-02-25 21:42:45Z - twisted.web2.xmlrpc
* [test_xmlrpc.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/081/081be7c370bf9e7b4c6e696276c1b2d57623b26b/a0ce57bf70d54841c52b964c628d9f98dcb3084f.py) (5838 bytes) - added by oubiwann on 2006-02-25 21:43:50Z - twisted.web2.test.test_xmlrpc

<details><summary>Searchable metadata</summary>

```
trac-id__1132 1132
type__enhancement enhancement
reporter__itamarst itamarst
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1123287285000000 1123287285000000
changetime__1142103387000000 1142103387000000
version__ 
owner__ 
cc__itamarst cc__jknight cc__oubiwann
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
1. Drop support for Handlers. They are useless.

2. xmlrpc_foo methods should always accept the Request, or whatever the
applicable web2 object is (context?) as their first argument; this should solve
one of the common questions people have about using XML-RPC.

3. Make exception handling more pluggable.

4. Make dispatch (_getFunction) public and pluggable.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

```
#!html
<pre>
As for #1, I'd vote to keep them. I use them all the time. They were 
actually one of the reasons I originally chose t.w.xmlrpc for my projects. 
Handlers help keep XML-RPC applicaiton APIs organized.

For what reason(s) do you want to drop Handlers?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

I talked with Itamar about this today, and he explained: you can still have the same functionality as what the XML-RPC handler provids by creating an object to do the same thing. For what it's worth, I agree with him now.

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

Itamar and I pair prog'ed today at pycon. He has added XML-RPC server functionality to t.web2. I will attach the files shortly.

We (he) finished [#11338](https://github.com/twisted/twisted/issues/11338), #2, and #4.

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

Additional notes: 
1) this is server-only (we're waiting for higher-level t.w2 http client code
2) the numbers 1,2 and 4 above refer to the number items from Itamar's first comment in this issue, and not to bug numbers.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Should I commit this, or do you want to make further changes first?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

I think it's ready to be committed.

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

I second that. I am working on some sample code/docs for t.w2.xmlrpc, but that will go in a separate ticket.

**Comment by twisted-trac** [comment]
|<img alt="foom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|foom set status to `closed`|
|-|-|

(In [16244]) Add itamar and oubiwann's xmlrpc server resource, with a couple mods.
Closes [#10744](https://github.com/twisted/twisted/issues/10744).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
